### DISCUSSION TITLE
Is there anyway to attach debugpy and lldb to same process ...?

### DISCUSSION THREAD

**Discussion by breathe** [microsoft/debugpy#1738] [Q&A]
I use `debugpy` from vscode to attach python debugger to a process and this [lldb extension](https://marketplace.visualstudio.com/items?itemName=vadimcn.vscode-lldb) to attach lldb to processes in order to debug code in cpp python extensions.

I have things setup so that for every entrypoint in my program I can choose to attach either debupy and debug the python portion of the execution or lldb and debug the cpp portion of the execution.

I would _really_ like to be able to attach both debuggers to the same process -- such that I could see the python call stack somehow when paused at a cpp breakpoint ...  Is there anyway to do that?  I've tried attaching both debuggers to the same process but it seems to lead to application hanging somewhere ...

**Comment by rchiodo** [comment]
The only way I know of to do this is to have a C++ debugger that understands Python frames and translates them into Python call stacks. 

Visual Studio 2022 does this with its [Python mixed mode debugging](https://learn.microsoft.com/en-us/visualstudio/python/debugging-mixed-mode-c-cpp-python-in-visual-studio?view=vs-2022). 

Otherwise, your LLDB debugger should be able to debug the python runtime itself. You'd have to do the stack translation yourself though.

For example in the VS 2022 mixed mode debugger, `_PyEval_EvalFrameDefault` indicates to the mixed mode debugger an entry into a python frame. You can evaluate the frame parameter passed to that function to figure out what python code is running (that's essentially what the debugger in VS 2022 does).

You can see the source for the VS 2022 mixed mode debugger here:
https://github.com/microsoft/PTVS/tree/main/Python/Product/Debugger.Concord

**Reply by rchiodo** [reply]
Attaching both debugpy and lldb to the same process might work, but they're fundamentally different. lldb is a hard mode debugger. It uses the system's debug events and Read/Write process memory to debug the process.

Debugpy is a soft mode debugger. It uses APIs built into Python to debug a python process. It's actually running inside your app. So stopping on breakpoints in LLDB can interfere with Debugpy being able to do things (as all threads are suspended).

**Comment by breathe** [comment]
Ok thanks very much for the quick reply and great info!

lldb has some suprising functionality in it via its (unwieldly) python lldb scripting api ... I wonder if I there's way to persuade lldb to interpret the python call frames ... even just to display the name of the last function in the python function callstack would be nice ...
